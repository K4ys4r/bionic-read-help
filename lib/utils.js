export const defaultSetting = {
  get bionicAlgorithm() {
    return [-1, 0.25, 0.25, 0.35, 0.5, 0.65];
  },
  get bionicTypes() {
    return { static: "static", interactive: "interactive" };
  },
  get bionicOpacity() {
    return 0.75;
  },
  get bionicTextColor() {
    return "#000";
  },
  get bionicStyles() {
    return {
      global: {
        display: "flex",
        "flex-direction": "column",
        border: "1px solid #a9a9a9",
        "border-radius": " 0.5rem",
        padding: "5px",
        overflow: "hidden",
        "text-align": "justify",
      },
      label: {
        display: "flex",
        "justify-content": "center",
        "align-items": "center",
        gap: "10px",
        "background-color": "#c5c5c5",
        color: "#000",
        padding: "5px 10px",
        "border-radius": "0.5rem",
        border: "1px solid #a0a0a0",
        "margin-top": "0.25rem",
        height: "1.5rem",
        "max-height": "2rem",
      },
      toolbar: {
        padding: "inherit",
        "background-color": "#d9d9d9",
        "margin-left": "-0.65rem",
        "margin-top": "-0.65rem",
        "margin-right": "-0.65rem",
        "margin-bottom": "0.75rem",
        display: "flex",
        "flex-wrap": "wrap",
        "flex-direction": "row",
        "justify-content": "space-around",
        "align-items": "center",
        "border-bottom": "1px solid #cecece",
      },
      input: {
        display: "flex",
        padding: "5px 0px",
        "text-align": "center",
        "border-radius": "0.25rem",
        "background-color": "#fff",
        border: "1px solid #a0a0a0",
      },
    };
  },
  get bionicToolsBar() {
    return {
      toolbar: {
        dom: () => {
          let bar = document.createElement("span");
          bar.id = "wc-bionic-toolbar";
          setElementStyleProperties(bar, defaultSetting.bionicStyles.toolbar);
          [
            defaultSetting.bionicToolsBar.bionic.dom(),
            defaultSetting.bionicToolsBar.opacity.dom(),
            defaultSetting.bionicToolsBar.algorithm.dom(),
            defaultSetting.bionicToolsBar["text-color"].dom(),
          ].map((dom) => bar.appendChild(dom));
          bar.setAttribute("part", bar.id);
          return bar;
        },
      },
      bionic: {
        id: "bionic-check",
        dom: () => {
          let label = document.createElement("label");
          label.textContent = "Bionic";
          let inp = document.createElement("input");
          inp.type = "checkbox";
          inp.id = defaultSetting.bionicToolsBar.bionic.id;
          setElementStyleProperties(inp, defaultSetting.bionicStyles.input);
          label.appendChild(inp);
          setElementStyleProperties(label, defaultSetting.bionicStyles.label);
          return label;
        },
      },
      opacity: {
        id: "bionic-opacity",
        dom: () => {
          let label = document.createElement("label");
          label.textContent = "Opacity";
          let inp = document.createElement("input");
          inp.id = defaultSetting.bionicToolsBar.opacity.id;
          inp.type = "range";
          inp.min = 0;
          inp.max = 1;
          inp.step = 0.01;
          setElementStyleProperties(inp, defaultSetting.bionicStyles.input);
          setElementStyleProperties(label, defaultSetting.bionicStyles.label);
          label.appendChild(inp);
          return label;
        },
      },
      "text-color": {
        id: "bionic-text-color",
        dom: () => {
          let label = document.createElement("label");
          label.textContent = "Text color";
          let inp = document.createElement("input");
          inp.id = defaultSetting.bionicToolsBar["text-color"].id;
          inp.type = "color";
          setElementStyleProperties(label, defaultSetting.bionicStyles.label);
          label.appendChild(inp);
          return label;
        },
      },
      algorithm: {
        id: "bionic-algorithm",
        dom: () => {
          let label = document.createElement("label");
          label.textContent = "Algorithm";
          let inp = document.createElement("input");
          inp.type = "text";
          inp.id = defaultSetting.bionicToolsBar.algorithm.id;
          setElementStyleProperties(inp, defaultSetting.bionicStyles.input);
          setElementStyleProperties(label, defaultSetting.bionicStyles.label);
          label.appendChild(inp);
          return label;
        },
      },
    };
  },
};

export function setElementStyleProperties(el, style) {
  for (const property in style) {
    el.style[property] = style[property];
  }
}

export function toggleAttribute(el, name) {
  el.toggleAttribute(name);
}
