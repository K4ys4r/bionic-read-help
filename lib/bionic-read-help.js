import { defaultSetting, setElementStyleProperties } from "./utils.js";

export default class BionicReadHelp extends HTMLElement {
  constructor() {
    super();
    this.content = "";
    this.isBionic = true;
    this.bionicOpacity = defaultSetting.bionicOpacity;
    this.bionicTextColor = defaultSetting.bionicTextColor;
    this.bionicAlgo = defaultSetting.bionicAlgorithm;
    this.bionicToolsBar = {
      dom: defaultSetting.bionicToolsBar.toolbar.dom(),
    };
    this.bionicType = defaultSetting.bionicTypes.static;
  }

  static get tagName() {
    return "wc-bionic-read-help";
  }

  static get attributes() {
    return Object.freeze({
      "brh-opacity": "brh-opacity",
      "brh-algorithm": "brh-algorithm",
      "brh-type": "brh-type",
      "brh-text-color": "brh-text-color",
    });
  }

  static get observedAttributes() {
    return Object.values(BionicReadHelp.attributes);
  }

  connectedCallback() {
    setTimeout(() => {
      this.content = this.#getContent();
      this.#render();
    });
  }

  attributeChangedCallback(name, oldValue, newValue) {
    switch (name) {
      case BionicReadHelp.attributes["brh-opacity"]:
        this.bionicOpacity = Number(newValue) >= 0 ? newValue : 1;
        break;
      case BionicReadHelp.attributes["brh-text-color"]:
        this.bionicTextColor = newValue;
        break;
      case BionicReadHelp.attributes["brh-algorithm"]:
        this.bionicAlgo = newValue
          ? Array.from(newValue.split(" ").map(Number).filter(Boolean))
          : defaultSetting.bionicAlgorithm;
        break;
      case BionicReadHelp.attributes["brh-type"]:
        this.bionicType =
          defaultSetting.bionicTypes[newValue] ||
          defaultSetting.bionicTypes.static;
        break;
    }
  }

  #render() {
    if (this.bionicType === defaultSetting.bionicTypes.interactive) {
      setElementStyleProperties(this, defaultSetting.bionicStyles.global);
    }
    this.#createToolsBar();
    this.updateElement();
  }

  #getContent() {
    return this.innerText
      .trim()
      .split(/\s+/g)
      .filter((a) => a.trim().length > 0);
  }

  updateElement() {
    // ToolsBar
    this.innerHTML = "";
    if (this.bionicType === defaultSetting.bionicTypes.interactive) {
      this.appendChild(this.bionicToolsBar.dom);
    }

    // content color
    this.style.color = this.bionicTextColor;

    // Content
    const textContent = this.isBionic
      ? this.#getBionicContent()
      : this.content.join(" ");
    const span = document.createElement("span");
    span.id = BionicReadHelp.tagName + "-text";
    span.innerHTML = textContent;
    this.appendChild(span);
  }

  #createToolsBar() {
    if (this.bionicType === defaultSetting.bionicTypes.static) return;
    // Get input by their Ids
    const bionicCheckBox = this.bionicToolsBar.dom.querySelector(
      `#${defaultSetting.bionicToolsBar.bionic.id}`
    );
    const inpOpacity = this.bionicToolsBar.dom.querySelector(
      `#${defaultSetting.bionicToolsBar.opacity.id}`
    );
    const inpAlgo = this.bionicToolsBar.dom.querySelector(
      `#${defaultSetting.bionicToolsBar.algorithm.id}`
    );
    const inpTextColor = this.bionicToolsBar.dom.querySelector(
      `#${defaultSetting.bionicToolsBar["text-color"].id}`
    );

    // set inputs events
    bionicCheckBox.checked = this.isBionic;
    bionicCheckBox.onchange = (e) => {
      this.isBionic = e.target.checked;
      inpAlgo.disabled = !this.isBionic;
      inpOpacity.disabled = !this.isBionic;
      inpTextColor.disabled = !this.isBionic;
      this.updateElement();
    };

    inpOpacity.value = this.bionicOpacity;
    inpOpacity.onchange = (e) => {
      this.bionicOpacity = e.target.value;
      this.updateElement();
    };
    inpTextColor.value = this.bionicTextColor;
    inpTextColor.oninput = (e) => {
      this.bionicTextColor = e.target.value;
      this.style.color = this.bionicTextColor;
    };

    // get numeric values from algorithm input data and update its value
    inpAlgo.value = this.bionicAlgo.join(" ");
    inpAlgo.onchange = (e) => {
      const algo = e.target.value;
      this.bionicAlgo = algo
        ? Array.from(algo.split(" ").map(Number).filter(Boolean))
        : defaultSetting.bionicAlgorithm;
      inpAlgo.value = this.bionicAlgo.join(" ");
      this.updateElement();
    };
  }

  #getBionicContent() {
    let bionicContent = [];
    for (let word of this.content) {
      bionicContent.push(this.#bionicWord(word));
    }
    return bionicContent.join(" ");
  }

  #bionicWord(w) {
    let algoCoef = this.bionicAlgo[w.length - 1] || this.bionicAlgo.at(-1);
    let boldChar = Math.ceil(w.length * algoCoef);
    return `<b>${w.slice(0, boldChar)}</b><span style='opacity:${
      this.bionicOpacity
    }'>${w.slice(boldChar)}</span>`;
  }
}

// Add the webComponent to customElements
customElements.define(BionicReadHelp.tagName, BionicReadHelp);
