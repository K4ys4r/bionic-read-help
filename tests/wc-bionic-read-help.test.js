import BionicReadHelp from "../lib/bionic-read-help";
import "../lib/bionic-read-help";
import { beforeEach, describe, expect, it } from "vitest";
import { createBionicHelpWC, delay } from "./utils";

describe("wc-bionic-read-help", () => {
  beforeEach(async () => {
    document.body.innerHTML = "";
  });
  it("add wc-bionic-read-help to the DOM mode interactive", async () => {
    let message = "Hello World";
    createBionicHelpWC({ message, type: "interactive" });
    await delay(1000);

    const wcBrh = document.querySelector(BionicReadHelp.tagName);
    expect(wcBrh).not.toBeNull();
    expect(wcBrh.textContent).toBe(message);
    expect(wcBrh.shadowRoot.innerHTML).toMatchSnapshot();

    // test bionic tool bar content
    const brhToolBar = wcBrh.shadowRoot.querySelector("#wc-bionic-toolbar");
    expect(brhToolBar).not.toBeNull();
    const opacityInp = brhToolBar.querySelector("#bionic-opacity");
    const algoInp = brhToolBar.querySelector("#bionic-algorithm ");
    const checkInp = brhToolBar.querySelector("#bionic-check");
    const textColorInp = brhToolBar.querySelector("#bionic-text-color");
    expect(opacityInp).not.toBeNull();
    expect(opacityInp.value).toBe("0.75");
    expect(algoInp).not.toBeNull();
    expect(algoInp.value.split(/\s/).map(Number)).toEqual([
      -1, 0.25, 0.25, 0.35, 0.5, 0.65,
    ]);
    expect(checkInp).not.toBeNull();
    expect(checkInp.checked).toEqual(true);
    expect(textColorInp).not.toBeNull();
    expect(textColorInp.value).toBe("#000000");
    expect(brhToolBar.innerHTML).toMatchSnapshot();

    // test bionic text content
    const brhText = wcBrh.shadowRoot.querySelector("#wc-bionic-read-help-text");
    const bolds = brhText.querySelectorAll("b");
    const spans = brhText.querySelectorAll("span");
    expect(brhText).not.toBeNull();
    expect(bolds.length).toBeGreaterThan(0);
    expect(spans.length).toBeGreaterThan(0);
    expect(brhText.innerHTML).toMatchSnapshot();
  });

  it("add wc-bionic-read-help to the DOM mode static", async () => {
    document.body.innerHTML =
      "<wc-bionic-read-help>Hello World</wc-bionic-read-help>";

    await delay(1000);

    const wcBrh = document.querySelector(BionicReadHelp.tagName);
    expect(wcBrh).not.toBeNull();
    expect(wcBrh.textContent).toBe("Hello World");
    expect(wcBrh.shadowRoot.innerHTML).toMatchSnapshot();

    // test bionic tool bar content
    const brhToolBar = wcBrh.shadowRoot.querySelector("#wc-bionic-toolbar");
    expect(brhToolBar).toBeNull();

    // test bionic text content
    const brhText = wcBrh.shadowRoot.querySelector("#wc-bionic-read-help-text");
    const bolds = brhText.querySelectorAll("b");
    const spans = brhText.querySelectorAll("span");
    expect(brhText).not.toBeNull();
    expect(bolds.length).toBeGreaterThan(0);
    expect(spans.length).toBeGreaterThan(0);
    expect(brhText.innerHTML).toMatchSnapshot();
  });
});
