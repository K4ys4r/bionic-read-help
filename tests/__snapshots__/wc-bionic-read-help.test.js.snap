// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`wc-bionic-read-help > add wc-bionic-read-help to the DOM mode interactive 1`] = `"<span id="wc-bionic-toolbar" style="padding: inherit; background-color: rgb(217, 217, 217); margin: -0.65rem -0.65rem 0.75rem -0.65rem; display: flex; flex-wrap: wrap; flex-direction: row; justify-content: space-around; align-items: center; border-bottom: 1px solid #cecece;" part="wc-bionic-toolbar"><label style="display: flex; justify-content: center; align-items: center; gap: 10px; background-color: rgb(197, 197, 197); color: rgb(0, 0, 0); padding: 5px 10px; border-radius: 0.5rem; border: 1px solid #a0a0a0; margin-top: 0.25rem; height: 1.5rem; max-height: 2rem;">Bionic<input type="checkbox" id="bionic-check" style="display: flex; padding: 5px 0px; text-align: center; border-radius: 0.25rem; background-color: rgb(255, 255, 255); border: 1px solid #a0a0a0;"></label><label style="display: flex; justify-content: center; align-items: center; gap: 10px; background-color: rgb(197, 197, 197); color: rgb(0, 0, 0); padding: 5px 10px; border-radius: 0.5rem; border: 1px solid #a0a0a0; margin-top: 0.25rem; height: 1.5rem; max-height: 2rem;">Opacity<input id="bionic-opacity" type="range" min="0" max="1" step="0.01" style="display: flex; padding: 5px 0px; text-align: center; border-radius: 0.25rem; background-color: rgb(255, 255, 255); border: 1px solid #a0a0a0;"></label><label style="display: flex; justify-content: center; align-items: center; gap: 10px; background-color: rgb(197, 197, 197); color: rgb(0, 0, 0); padding: 5px 10px; border-radius: 0.5rem; border: 1px solid #a0a0a0; margin-top: 0.25rem; height: 1.5rem; max-height: 2rem;">Algorithm<input type="text" id="bionic-algorithm" style="display: flex; padding: 5px 0px; text-align: center; border-radius: 0.25rem; background-color: rgb(255, 255, 255); border: 1px solid #a0a0a0;"></label><label style="display: flex; justify-content: center; align-items: center; gap: 10px; background-color: rgb(197, 197, 197); color: rgb(0, 0, 0); padding: 5px 10px; border-radius: 0.5rem; border: 1px solid #a0a0a0; margin-top: 0.25rem; height: 1.5rem; max-height: 2rem;">Text color<input id="bionic-text-color" type="color"></label></span><span id="wc-bionic-read-help-text" part="wc-bionic-read-help-text"><b>Hel</b><span style="opacity:0.75">lo</span> <b>Wor</b><span style="opacity:0.75">ld</span></span>"`;

exports[`wc-bionic-read-help > add wc-bionic-read-help to the DOM mode interactive 2`] = `"<label style="display: flex; justify-content: center; align-items: center; gap: 10px; background-color: rgb(197, 197, 197); color: rgb(0, 0, 0); padding: 5px 10px; border-radius: 0.5rem; border: 1px solid #a0a0a0; margin-top: 0.25rem; height: 1.5rem; max-height: 2rem;">Bionic<input type="checkbox" id="bionic-check" style="display: flex; padding: 5px 0px; text-align: center; border-radius: 0.25rem; background-color: rgb(255, 255, 255); border: 1px solid #a0a0a0;"></label><label style="display: flex; justify-content: center; align-items: center; gap: 10px; background-color: rgb(197, 197, 197); color: rgb(0, 0, 0); padding: 5px 10px; border-radius: 0.5rem; border: 1px solid #a0a0a0; margin-top: 0.25rem; height: 1.5rem; max-height: 2rem;">Opacity<input id="bionic-opacity" type="range" min="0" max="1" step="0.01" style="display: flex; padding: 5px 0px; text-align: center; border-radius: 0.25rem; background-color: rgb(255, 255, 255); border: 1px solid #a0a0a0;"></label><label style="display: flex; justify-content: center; align-items: center; gap: 10px; background-color: rgb(197, 197, 197); color: rgb(0, 0, 0); padding: 5px 10px; border-radius: 0.5rem; border: 1px solid #a0a0a0; margin-top: 0.25rem; height: 1.5rem; max-height: 2rem;">Algorithm<input type="text" id="bionic-algorithm" style="display: flex; padding: 5px 0px; text-align: center; border-radius: 0.25rem; background-color: rgb(255, 255, 255); border: 1px solid #a0a0a0;"></label><label style="display: flex; justify-content: center; align-items: center; gap: 10px; background-color: rgb(197, 197, 197); color: rgb(0, 0, 0); padding: 5px 10px; border-radius: 0.5rem; border: 1px solid #a0a0a0; margin-top: 0.25rem; height: 1.5rem; max-height: 2rem;">Text color<input id="bionic-text-color" type="color"></label>"`;

exports[`wc-bionic-read-help > add wc-bionic-read-help to the DOM mode interactive 3`] = `"<b>Hel</b><span style="opacity:0.75">lo</span> <b>Wor</b><span style="opacity:0.75">ld</span>"`;

exports[`wc-bionic-read-help > add wc-bionic-read-help to the DOM mode static 1`] = `"<span id="wc-bionic-read-help-text" part="wc-bionic-read-help-text"><b>Hel</b><span style="opacity:0.75">lo</span> <b>Wor</b><span style="opacity:0.75">ld</span></span>"`;

exports[`wc-bionic-read-help > add wc-bionic-read-help to the DOM mode static 2`] = `"<b>Hel</b><span style="opacity:0.75">lo</span> <b>Wor</b><span style="opacity:0.75">ld</span>"`;
